Implemented Pooling kernel for CMSIS-NN backend

lamarrr opened the pull request:
Highlight:
- Implemented average and max pooling kernel for cmsis-nn backend
- Implemented reshape kernels for cmsis-nn backend
- Encapsulated scratch buffer

lamarrr committed:
Merge pull request #1 from tensorflow/master

merge

lamarrr committed:
Implemented average and max pooling kernel for cmsis-nn backend.

A few Notes:
	- The raw backend doesn't utilize EXPECTs but the aborting DCHECKs only in the reference_ops scope
	- The raw backend kernel's AverageEvalInt8 doesn't return a TfLiteStatus, but I added that to the cmsis-nn AverageEvalInt8 implementation since cmsis-nn returns status codes
	- The max pooling kernel on the raw backend doesn't have any implementation for the Int8 data type, so I haven't implemented one for the cmsis-nn backend either
	- I also utilize freddan80's temporary buffer allocator (as noted in PR #32867)

lamarrr committed:
Implemented reshape kernels for cmsis-nn backend

It has the same implementation as the raw backend.
It is basically just a copy operation, same as the cmsis-nn backend.

lamarrr committed:
Encapsulated scratch buffer

I removed the buffer from global scope. Which I think makes more sense in case of future changes.

lamarrr committed:
Merge pull request #2 from tensorflow/master

Merge tensorflow contributions

freddan80 commented on tensorflow/lite/experimental/micro/kernels/cmsis-nn/pooling.cc:
```diff
@@ -0,0 +1,278 @@
+                     scratch_buffer, GetTensorData<int8_t>(output)),
+      ARM_MATH_SUCCESS);
+#else
+#error ARM_MATH_DSP and ARM_MATH_LOOPUNROLL must be set
```
I know this is what we do for the other opt kernels. However, we're going to change this pattern to fall back to the reference kernel implementation. So something like:

#else
  tflite::reference_ops::AveragePool(...)
#endif

It would be great if you added that!

lamarrr replied:
Thanks for the info!
I have made the necessary changes

freddan80 commented on tensorflow/lite/experimental/micro/kernels/cmsis-nn/reshape.cc:
```diff
@@ -0,0 +1,100 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
```
This file is not needed if it doesn't call arm_reshape_s8(). 

See: 
https://www.tensorflow.org/lite/microcontrollers/library#optimized_kernels
for more details. If you remove this function, the build system will fallback to use
'kernels/reshape.cc'
instead of 
'kernels/cmsis-nn/reshape.cc'

gmiodice replied:
Since reshape is just a memcpy, I would suggest not integrating this function from CMSIS-NN because it does not bring any performance gain.

lamarrr replied:
> This file (cmsis-nn/reshape.cc) is not needed if it doesn't call arm_reshape_s8().
> 
> See:
> https://www.tensorflow.org/lite/microcontrollers/library#optimized_kernels
> for more details. If you remove this function, the build system will fallback to use
> 'kernels/reshape.cc'
> instead of
> 'kernels/cmsis-nn/reshape.cc'

I have done that and also added the reference implementation fallback you suggested. 
Should I add the fallback for the ones you implemented? (_conv, depthwise_conv, and fully_connected_)

lamarrr replied:
> Since reshape is just a memcpy, I would suggest not integrating this function from CMSIS-NN because it does not bring any performance gain.


I thought in that direction too.
But thinking about it now, I think it makes more sense to use _arm_reshape_s8_ since ARM can change the implementation anytime irregardless of whether it's just a plain _memcpy_.

freddan80 replied:
> Since reshape is just a memcpy, I would suggest not integrating this function from CMSIS-NN because it does not bring any performance gain.

@lamarrr Considering that this function doesn't give a performance improvement, I think you can remove this file (kernels/cmsis-nn/reshape.cc).

freddan80 replied:
> > This file (cmsis-nn/reshape.cc) is not needed if it doesn't call arm_reshape_s8().
> > See:
> > https://www.tensorflow.org/lite/microcontrollers/library#optimized_kernels
> > for more details. If you remove this function, the build system will fallback to use
> > 'kernels/reshape.cc'
> > instead of
> > 'kernels/cmsis-nn/reshape.cc'
> 
> I have done that and also added the reference implementation fallback you suggested.
> Should I add the fallback for the ones you implemented? (_conv, depthwise_conv, and fully_connected_)

I wouldn't mind at all ;) Thanks!

freddan80 replied:
> > Since reshape is just a memcpy, I would suggest not integrating this function from CMSIS-NN because it does not bring any performance gain.
> 
> I thought in that direction too.
> But thinking about it now, I think it makes more sense to use _arm_reshape_s8_ since ARM can change the implementation anytime irregardless of whether it's just a plain _memcpy_.

You're exactly correct in the way you think. Me and @gmiodice represent ARM ;) we don't see optimization potential there currently - so we're discussing to remove it. @gmiodice , correct me if I'm wrong.

gmiodice replied:
Hi both, yes I confirm that since reshape does not bring any performance gain, we could remove that function from CMSIS-NN. Thanks for checking @freddan80 and thanks @lamarrr for your contribution ;)

lamarrr replied:
> > > Since reshape is just a memcpy, I would suggest not integrating this function from CMSIS-NN because it does not bring any performance gain.
> > 
> > 
> > I thought in that direction too.
> > But thinking about it now, I think it makes more sense to use _arm_reshape_s8_ since ARM can change the implementation anytime irregardless of whether it's just a plain _memcpy_.
> 
> You're exactly correct in the way you think. Me and @gmiodice represent ARM ;) we don't see optimization potential there currently - so we're discussing to remove it. @gmiodice , correct me if I'm wrong.

Great!
Alright, I'll be awaiting your conclusion.

lamarrr replied:
> Hi both, yes I confirm that since reshape does not bring any performance gain, we could remove that function from CMSIS-NN. Thanks for checking @freddan80 and thanks @lamarrr for your contribution ;)

My pleasure

freddan80 replied:
> So... I should remove the _cmsis-nn/reshape.cc_ source file?

yes

lamarrr replied:
> yes

ok. will do

lamarrr replied:
> > yes
> 
> ok. will do

Done

freddan80 reviewed (changes requested):
Thanks for your contributions @lamarrr ! I have added some comments

lamarrr committed:
Added reference average pooling kernel fallback

ensured the kernel calls the raw tensorflow api instead of the compile time error

lamarrr committed:
Changed raw reshape kernel's copy to the CMSIS-NN style reshape

Changed the reshape kernel's loop-based copy to the cmsis-nn style reshape

lamarrr committed:
Added reference kernel fallback

Added reference kernel fallback to the reshape op. arm_reshape_s8 is a memcpy though there's little difference

freddan80 commented on tensorflow/lite/experimental/micro/kernels/cmsis-nn/scratch_buffer.cc:
```diff
@@ -15,20 +15,21 @@ limitations under the License.
+  // network.
+
+  __attribute__((aligned(
+      4))) static int16_t cmsis_scratch_buffer[SCRATCH_BUFFER_BYTES / 2] = {0};
```
@lamarrr there's another PR from you fixing this as well. Just a heads up to avoid merge conflicts...

lamarrr replied:
I'll close that. Thanks

lamarrr replied:
@freddan80, I've added the kernel fallbacks. Please do a review and give feedback  ;-)

lamarrr committed:
Removed reshape kernel

removed reshape kernel implementation, since this does not bring any performance gain from the cmsis-nn backend

freddan80 commented:
@lamarrr Nice job! Let me know if you need input on the fallback solutions for the other cmsis-nn kernels. Or anything else regarding optimization. Cheers!

lamarrr commented:
> @lamarrr Nice job! Let me know if you need input on the fallback solutions for the other cmsis-nn kernels. Or anything else regarding optimization. Cheers!

@freddan80  
Ok, sir. Will do!

lamarrr committed:
Merge branch 'master' into master

lamarrr commented:
> @lamarrr Nice job! Let me know if you need input on the fallback solutions for the other cmsis-nn kernels. Or anything else regarding optimization. Cheers!

@freddan80 
any plans to implement _broadcasting add_ for the cmsis-nn _add_ kernel?
I think that's the only thing left for the tflite _add_  kernel.

freddan80 commented:
> > @lamarrr Nice job! Let me know if you need input on the fallback solutions for the other cmsis-nn kernels. Or anything else regarding optimization. Cheers!
> 
> @freddan80
> any plans to implement _broadcasting add_ for the cmsis-nn _add_ kernel?
> I think that's the only thing left for the tflite _add_ kernel.

It's on our todo-list, but not until beginning of next year. If you want to contribute it earlier, we'd be happy to help out. Cheers! /Fredrik
